File: app/src/main/java/com/exzell/mangaplayground/io/Repository.java
package com.exzell.mangaplayground.io;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.exzell.mangaplayground.AppExecutors;
import com.exzell.mangaplayground.download.Download;
import com.exzell.mangaplayground.io.database.AppDatabase;
import com.exzell.mangaplayground.io.database.ChapterDao;
import com.exzell.mangaplayground.io.database.DBManga;
import com.exzell.mangaplayground.io.database.DownloadDao;
import com.exzell.mangaplayground.io.database.MangaDao;
import com.exzell.mangaplayground.io.internet.InternetManager;
import com.exzell.mangaplayground.io.internet.MangaParkApi;
import com.exzell.mangaplayground.models.Chapter;
import com.exzell.mangaplayground.models.Manga;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.rxjava3.core.Observable;
import okhttp3.Request;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;

@Singleton
public class Repository {

    private AppExecutors mExecutor;
    private MangaParkApi mMangaPark;

    private ChapterDao mChapterDao;
    private MangaDao mMangaDao;
    private DownloadDao mDownloadDao;

    @Inject
    public Repository(AppExecutors exec, Retrofit service, AppDatabase db) {
        mExecutor = exec;

        mMangaPark = service.create(MangaParkApi.class);

        mMangaDao = db.getMangaDao();
        mChapterDao = db.getChapterDao();
        mDownloadDao = db.getDownloadDao();
    }

    //TODO: Consider changing the return type of both methods to Observables

    public Call<ResponseBody> advancedSearch(Map<String, String> queries) {

        try {
            return mExecutor.getIoExecutor().submit(() -> mMangaPark.advancedSearch(queries)).get();
        } catch (CancellationException | ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    public Response<ResponseBody> home() {
        try {
            return mExecutor.getIoExecutor().submit(() -> mMangaPark.home().execute()).get();
        } catch (CancellationException | ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * Moves to a particular page in the same website
     */
    public Observable<ResponseBody> moveTo(String link) { return mMangaPark.next(link); }

    /**
     * Moves to an entirely different website
     */
    public Observable<ResponseBody> goTo(String link) {
        try {
            Request req = new Request.Builder().url(link).build();
            return Observable.just(InternetManager.mClient.newCall(req).execute().body());
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }


    //Database calls

    //Calls to MangaDao

    /**
     * Takes care of inserting the mangas into the DB. After the manga id is gotten from the DB,
     * it is set into its manga instance, so be sure not to replace the instances passed to this method
     */
    public void insertManga(Manga... manga) {
        mExecutor.getDiskExecutor().submit((() -> {
            for (Manga man : manga) {
                long id = mMangaDao.insertMangas(man);
                man.getChapters().forEach(chap -> chap.setMangaId(id));
                man.setId(id);
            }

            mChapterDao.insertChapters(Arrays.stream(manga).flatMap(man -> man.getChapters()
                    .stream()).collect(Collectors.toList()));
        }));
    }

    public void updateManga(boolean andChapters, Manga... manga) {
        mExecutor.getDiskExecutor().submit(() -> {
            mMangaDao.updateMangas(Arrays.asList(manga));
            if (andChapters) mChapterDao.insertChapters(Arrays.stream(manga)
                    .flatMap(man -> man.getChapters().stream()).collect(Collectors.toList()));
        });
    }

    public Manga getMangaWithLink(String link) {
        try {
            return mExecutor.getDiskExecutor().submit((Callable<Manga>) () -> mMangaDao.getMangaFromLink(link)).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

//    public List<Manga> getMangas(){
//        List<Manga> manga = new ArrayList<>();
//
//        try {
//            return mExecutor.getDiskExecutor().submit(() -> mMangaDao.getMangas()).get();
//        } catch (ExecutionException | InterruptedException e) {
//            e.printStackTrace();
//            return Collections.emptyList();
//        }
//    }

    public List<DBManga> getBookmarkedMangaNotLive() {

        try {
            return mExecutor.getDiskExecutor().submit(() -> mMangaDao.notLiveBookmarks()).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return Collections.emptyList();
        }
    }

    public LiveData<List<DBManga>> getBookmarkedManga() {
        try {
            return mExecutor.getDiskExecutor().submit(() -> mMangaDao.bookmarks()).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    public LiveData<List<DBManga>> getDownloadedMangas() {
        try {
            return mExecutor.getDiskExecutor().submit(() -> mMangaDao.downloads()).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    @NotNull
    public List<DBManga> getMangaWithIds(@NonNull List<Long> ids) {

        try {
            return mExecutor.getDiskExecutor().submit(() -> ids.stream().map(s ->
                    mMangaDao.getMangaFromId(s)).collect(Collectors.toList())).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return Collections.emptyList();
        }
    }

    public List<DBManga> lastReadMangas(long time) {
        try {
            return mExecutor.getDiskExecutor().submit(() -> mMangaDao.getMangaFromChapterTime(time)).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return Collections.emptyList();
        }
    }

    public DBManga getMangaForChapter(long id) {

        try {
            return mExecutor.getDiskExecutor().submit(() -> mMangaDao.getMangaFromChapter(id)).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    //Calls to ChapterDao

    public void updateChapters(List<Chapter> chapter) {
        mExecutor.getDiskExecutor().submit(() -> mChapterDao.updateChapters(chapter));
    }

    /**
     * Returns the timestamp for every manga with atleast 1 chapter with a read time greater than 0
     */
    public LiveData<List<Long>> allMangaTime() {
        try {
            return mExecutor.getDiskExecutor().submit(() -> mChapterDao.allMangaTime()).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    //Calls to DownloadDao

    public void insertDownloads(List<Download> downs) {
        new Thread(() -> mDownloadDao.addDownloads(downs)).start();
    }

    public void deleteDownloads(List<Download> downs) {
        new Thread(() -> mDownloadDao.deleteDownloads()).start();
    }

    public LiveData<List<Download>> getLiveDownloads() {

        try {
            return mExecutor.getDiskExecutor().submit(() -> mDownloadDao.getPendingDownloadsLive()).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    public LiveData<List<Download>> getDownloads() {
        try {
            return mExecutor.getDiskExecutor().submit(() -> mDownloadDao.getAllDownloads()).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    public List<Download> getCurrentDownloads() {

        try {
            return mExecutor.getDiskExecutor().submit(() -> mDownloadDao.getPendingDownloads()).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }

    public void updateDownloads(List<Download> d) {
        mExecutor.getDiskExecutor().submit(() -> mDownloadDao.updateDownloads(d));
    }

    public String getDownloadPath(long chapterId) {
        try {
            return mExecutor.getDiskExecutor().submit(() -> mDownloadDao.getPathFromId(chapterId)).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/java/com/exzell/mangaplayground/io/Repository.kt
package com.exzell.mangaplayground.io

import com.exzell.mangaplayground.AppExecutors
import com.exzell.mangaplayground.download.model.DownloadManga
import com.exzell.mangaplayground.io.database.*
import com.exzell.mangaplayground.io.internet.InternetManager
import com.exzell.mangaplayground.io.internet.MangaParkApi
import com.exzell.mangaplayground.models.Chapter
import com.exzell.mangaplayground.models.Download
import com.exzell.mangaplayground.models.Manga
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers
import kotlinx.coroutines.flow.Flow
import okhttp3.Request
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.Retrofit
import java.io.IOException
import java.util.concurrent.Callable
import java.util.concurrent.CancellationException
import java.util.concurrent.ExecutionException
import java.util.function.Consumer
import java.util.stream.Collectors
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class Repository @Inject constructor(private val mExecutor: AppExecutors, service: Retrofit, db: AppDatabase) {
    private val mMangaPark: MangaParkApi = service.create(MangaParkApi::class.java)
    private val mChapterDao: ChapterDao = db.chapterDao
    private val mMangaDao: MangaDao = db.mangaDao
    private val mDownloadDao: DownloadDao = db.downloadDao


    //TODO: Consider changing the return type of both methods to Observables
    fun advancedSearch(queries: Map<String, String>): Observable<ResponseBody>? {
        return try {
            mMangaPark.advancedSearch(queries).subscribeOn(Schedulers.io())
        } catch (e: CancellationException) {
            e.printStackTrace()
            null
        } catch (e: ExecutionException) {
            e.printStackTrace()
            null
        } catch (e: InterruptedException) {
            e.printStackTrace()
            null
        }
    }

    fun home(): Response<ResponseBody>? {
        return try {
            mExecutor.ioExecutor.submit(Callable { mMangaPark.home().execute() }).get()
        } catch (e: CancellationException) {
            e.printStackTrace()
            null
        } catch (e: ExecutionException) {
            e.printStackTrace()
            null
        } catch (e: InterruptedException) {
            e.printStackTrace()
            null
        }
    }

    /**
     * Moves to a particular page in the same website
     */
    fun moveTo(link: String): Observable<ResponseBody> {
        return mMangaPark.next(link)
    }

    /**
     * Moves to an entirely different website
     */
    fun goTo(link: String): Observable<ResponseBody>? {
        return try {
            val req = Request.Builder().url(link).build()
            Observable.just(InternetManager.client.newCall(req).execute().body())
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }
    //Database calls
    //Calls to MangaDao
    /**
     * Takes care of inserting the mangas into the DB. After the manga id is gotten from the DB,
     * it is set into its manga instance, so be sure not to replace the instances passed to this method
     */
    fun insertManga(vararg manga: Manga) {
        mExecutor.diskExecutor.submit {
            for (man in manga) {
                val id = mMangaDao.insertMangas(man)
                man.chapters.forEach(Consumer { chap: Chapter -> chap.mangaId = id })
                man.id = id
            }
            mChapterDao.insertChapters(manga.flatMap {
                it.chapters
            })
        }
    }

    fun updateManga(andChapters: Boolean, vararg manga: Manga) {
        mExecutor.diskExecutor.submit {
            mMangaDao.updateMangas(listOf(*manga))

            if (!andChapters) return@submit

            //separate new from old: key 1 for those with ids(i.e already in the DB) and 2 for those without one
            val chapterGroup = manga.flatMap {
                it.chapters.also { chapLst ->
                    chapLst.filter { ch -> ch.mangaId <= 0 }
                            .onEach { ch -> ch.mangaId = it.id }
                }
            }.groupBy { chap -> if (chap.id > 0) 1 else 2 }

            chapterGroup.keys.forEach {
                chapterGroup[it]?.let { group ->
                    if (it == 1) mChapterDao.updateChapters(group)
                    else mChapterDao.insertChapters(group)
                }
            }
        }
    }

    fun updateManga(vararg infos: BookmarkInfo) {
        mExecutor.diskExecutor.submit {
            mMangaDao.updateMangasPart(listOf(*infos))
        }
    }

    fun getMangaWithLink(link: String): Manga? {
        return try {
            mExecutor.diskExecutor.submit(Callable<Manga> { mMangaDao.getMangaFromLink(link) }).get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
            null
        } catch (e: InterruptedException) {
            e.printStackTrace()
            null
        }
    }

    fun getBookmarkedMangaNotLive(): List<DBManga> = try {
        mExecutor.diskExecutor.submit(Callable { mMangaDao.notLiveBookmarks() }).get()
    } catch (e: ExecutionException) {
        e.printStackTrace()
        emptyList()
    } catch (e: InterruptedException) {
        e.printStackTrace()
        emptyList()
    }

    fun getBookmarkedManga(): Flow<List<BookmarkInfo>> {
        return mMangaDao.bookmarks()
    }

    fun getDownloadedMangas(): Flow<List<BookmarkInfo>> {
        return mMangaDao.downloads()
    }

    fun getMangaWithIds(ids: List<Long?>): List<DBManga> {
        return try {
            mExecutor.diskExecutor.submit(Callable { ids.stream().map { s: Long? -> mMangaDao.getMangaFromId(s!!) }.collect(Collectors.toList()) }).get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
            emptyList()
        } catch (e: InterruptedException) {
            e.printStackTrace()
            emptyList()
        }
    }

    fun lastReadMangas(time: Long): List<HistoryInfo> {
        return try {
            mExecutor.diskExecutor.submit(Callable { mMangaDao.lastReadInfo(time).distinctBy { it.id } }).get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
            emptyList()
        } catch (e: InterruptedException) {
            e.printStackTrace()
            emptyList()
        }
    }

    fun getMangaForChapter(id: Long): DBManga? {
        return try {
            mExecutor.diskExecutor.submit(Callable { mMangaDao.getMangaFromChapter(id) }).get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
            null
        } catch (e: InterruptedException) {
            e.printStackTrace()
            null
        }
    }

    fun getDownloadManga(chapterId: Long): DownloadManga? {
        return try {
            mExecutor.diskExecutor.submit(Callable { mMangaDao.getMangaInfoFromId(chapterId).createManga() }).get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
            null
        } catch (e: InterruptedException) {
            e.printStackTrace()
            null
        }
    }

    //Calls to ChapterDao
    fun updateChapters(chapter: List<Chapter>) {
        mExecutor.diskExecutor.submit { mChapterDao.updateChapters(chapter) }
    }

    fun updateChapterTime(chapters: List<ChapterTimeUpdate>) {
        mExecutor.diskExecutor.submit { mChapterDao.updateChaptersTime(chapters) }
    }

    /**
     * Returns the timestamp for every manga with atleast 1 chapter with a read time greater than 0
     */
    fun allMangaTime(): Flow<List<Long>> {
        return mChapterDao.allMangaTime()
    }

    //Calls to DownloadDao
    fun insertDownloads(downs: List<Download>) {
        mExecutor.diskExecutor.submit { mDownloadDao.addDownloads(downs) }
    }

    fun deleteDownloads(downs: List<Download>) {
        mExecutor.diskExecutor.submit { mDownloadDao.deleteDownloads(downs) }
    }

    /** Returns the list of all uncompleted downloads */
    fun getLiveDownloads(): Flow<List<Download>> {
        return mDownloadDao.getPendingDownloadsLive()
    }

    /** Returns a list of every download in the db*/
    fun getDownloads(): Flow<List<Download>> {
        return mDownloadDao.getAllDownloads()
    }

    /** Returns a list of chapter_ids for all the completed downloads **/
    fun getCompletedDownloadChapterIds(): Flow<List<Long>> {
        return mDownloadDao.getCompleteIds()
    }

    fun getCurrentDownloads(): List<Download> = try {
        mExecutor.diskExecutor.submit(Callable { mDownloadDao.getPendingDownloads() }).get()
    } catch (e: ExecutionException) {
        e.printStackTrace()
        emptyList()
    } catch (e: InterruptedException) {
        e.printStackTrace()
        emptyList()
    }

    fun updateDownloads(d: List<Download>) {
        mExecutor.diskExecutor.submit { mDownloadDao.updateDownloads(d) }
    }

    fun getDownloadPath(chapterId: Long): String? {
        return try {
            mExecutor.diskExecutor.submit<String> { mDownloadDao.getPathFromId(chapterId) }.get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
            null
        } catch (e: InterruptedException) {
            e.printStackTrace()
            null
        }
    }
}